source/opt/local_ssa_elim_pass.cpp

```diff
@@ -0,0 +1,759 @@
+  const auto uitr = type2undefs_.find(type_id);
+  if (uitr != type2undefs_.end())
+    return uitr->second;
+  const uint32_t undefId = TakeNextId();
+  std::unique_ptr<ir::Instruction> undef_inst(
+    new ir::Instruction(SpvOpUndef, type_id, undefId, {}));
+  def_use_mgr_->AnalyzeInstDefUse(&*undef_inst);
+  module_->AddGlobalValue(std::move(undef_inst));
+  type2undefs_[type_id] = undefId;
+  return undefId;
+}
+
+uint32_t LocalSSAElimPass::GetPointeeTypeId(
+    const ir::Instruction* ptrInst) const {
+  const uint32_t ptrTypeId = ptrInst->type_id();
+  const ir::Instruction* ptrTypeInst = def_use_mgr_->GetDef(ptrTypeId);
+  return ptrTypeInst->GetSingleWordInOperand(kTypePointerTypeIdInIdx);
+}
+
+void LocalSSAElimPass::SSABlockInitLoopHeader(
+    std::list<ir::BasicBlock*>::iterator block_itr) {
+  const uint32_t label = (*block_itr)->id();
+  // Determine backedge label.
+  uint32_t backLabel = 0;
+  for (uint32_t predLabel : label2preds_[label])
+    if (visitedBlocks_.find(predLabel) == visitedBlocks_.end()) {
+      assert(backLabel == 0);
+      backLabel = predLabel;
+      break;
+    }
+  assert(backLabel != 0);
+  // Determine merge block.
+  auto mergeInst = (*block_itr)->begin();
+  uint32_t mergeLabel = mergeInst->GetSingleWordInOperand(
+      kLoopMergeMergeBlockIdInIdx);
+  // Collect all live variables and a default value for each across all
+  // non-backedge predecesors. Must be ordered map because phis are
+  // generated based on order and test results will otherwise vary across
+  // platforms.
+  std::map<uint32_t, uint32_t> liveVars;
+  for (uint32_t predLabel : label2preds_[label]) {
+    for (auto var_val : label2ssa_map_[predLabel]) {
+      uint32_t varId = var_val.first;
+      liveVars[varId] = var_val.second;
+    }
+  }
+  // Add all stored variables in loop. Set their default value id to zero.
+  for (auto bi = block_itr; (*bi)->id() != mergeLabel; ++bi) {
+    ir::BasicBlock* bp = *bi;
+    for (auto ii = bp->begin(); ii != bp->end(); ++ii) {
+      if (ii->opcode() != SpvOpStore)
+        continue;
+      uint32_t varId;
+      (void) GetPtr(&*ii, &varId);
+      if (!IsTargetVar(varId))
+        break;
+      liveVars[varId] = 0;
+    }
+  }
+  // Insert phi for all live variables that require them. All variables
+  // defined in loop require a phi. Otherwise all variables
+  // with differing predecessor values require a phi.
+  auto insertItr = (*block_itr)->begin();
+  for (auto var_val : liveVars) {
+    const uint32_t varId = var_val.first;
+    if (!IsLiveAfter(varId, label))
+      continue;
+    const uint32_t val0Id = var_val.second;
+    bool needsPhi = false;
+    if (val0Id != 0) {
+      for (uint32_t predLabel : label2preds_[label]) {
+        // Skip back edge predecessor.
+        if (predLabel == backLabel)
+          continue;
+        const auto var_val_itr = label2ssa_map_[predLabel].find(varId);
+        // Missing values do not cause difference
```

Comment:
Why do missing values not cause a difference when processing loops, but they do cause a difference in the multi-pred case?   Should explain why we consider no difference here.  Is this a bug?

Example:
Let's say this header H has two predecessors P and Q; P does not dominate Q; that variable V has value p0 when coming from P, that p0 is defined in P, and V has no value when coming from Q.  This code says no difference needs to be resolved.   Then let's say the loop headed by H loads from V.  We're trying to replace that load with an SSA value, and I think this code will replace it with p0.  But that use of p0 in the loop is not dominated by the definition in P since we could have reached H via Q instead of via P.